### DISCUSSION TITLE
TypeError: traverse is not a function

### DISCUSSION THREAD

**Discussion by khushal87** [babel/babel#13093] [Q&A] [answered]
My code snippet - 
```
import babel from "@babel/parser";
import traverse from "@babel/traverse";

const code = `
const x : number  = 10;
`; 

const ast = babel.parse(code, {
    plugins: ["typescript"],
});

traverse(ast, {
    enter(path) {
        if (path.node.type == "Identifier") {
            console.log(path);
        }
    },
});

console.log(ast.type);

```

I am continuously getting:
```
traverse(ast, {
^

TypeError: traverse is not a function
```

Can anyone help me out. I have used the package @babel/traverse as required to use the functionality but I don't find why isn't this working. Any help would be great. Thanks

**Comment by fedeci** [comment]
You are using `typescript` with `ESModules`, right?

**Reply by khushal87** [reply]
Yes I am. Can you help how should I use in this case? @fedeci

**Comment by nicolo-ribaudo** [comment, marked as answer]
If you are using native ESM, you need to use `traverse.default` instead of `traverse` (that's how Node.js' interop between ESM and CJS works).

**Reply by khushal87** [reply]
Can you give an example? How should I import in this case? @nicolo-ribaudo

**Reply by nicolo-ribaudo** [reply]
```js
import traverse from "@babel/traverse";

traverse.default(ast, { ... })
```

**Reply by khushal87** [reply]
This worked like charm. Thanks @nicolo-ribaudo :)

**Comment by largis21** [comment]
I ended up creating a file for exporting traverse like this, because Vitest actually get the default export automaticly
```ts
import _traverse from "@babel/traverse";

export const traverse = process.env["TEST"]
  ? _traverse
  : // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ((_traverse as any).default as typeof _traverse);
```

**Reply by josefmic** [reply]
Hey, how did you get this to work? I still get "TypeError: traversae is not a function" when the file is built. If I run the file locally it works fine.

**Reply by nicolo-ribaudo** [reply]
Use `typeof _traverse === "function" ? _traverse : _traverse.default`

**Reply by josefmic** [reply]
Should've figured... Thank you, you are a lifesaver!

**Comment by morganney** [comment]
An issue here is if you write ESM with typescript, and then want to run a dual build with `tsc`. You'll get `_traverse.default.default` emitted ,so you're back in the `TypeError: traverse is not a function` realm.

Default imports are more trouble than they're worth.

I circumvented this issue with:

```ts
import { moduleType } from 'node-module-type'

const type = moduleType()
const traverse = (
  typeof _traverse === 'function' || type === 'commonjs' ? _traverse : _traverse.default
) as typeof _traverse.default
```

Now you can test and dual build.

**Comment by zlobz** [comment]
```js
import _traverse from "@babel/traverse";
const traverse = _traverse.default;
```
